src/app-layer-detect-proto.c

```diff
@@ -345,7 +345,9 @@ static AppProto AppLayerProtoDetectPPGetProto(Flow *f,
             SCLogDebug("toserver - Probing parser found for source port %"PRIu16, f->sp);
 
             /* found based on source port, so use sp registration */
-            pe2 = pp_port_sp->sp;
+            pe2 = pp_port_sp->dp;
+            StreamTcpPacketSwitchDir(NULL, p);
```

Comment:
I see the problem of the applayer recognizing that the flow's direction is reversed, but we'll need a better solution for dealing with that.